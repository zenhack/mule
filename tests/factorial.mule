#XFAIL: match not implemented
let factorial =
  fn n. match n with
  | 0 -> 1
  | n -> int.mul n (factorial (int.sub n 1))
  end
in
factorial 5
