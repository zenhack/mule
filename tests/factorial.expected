
Desugared:

	(letrec (types ())
	 (values
	  ((factorial
	    (fn n
	     ((match
	       ((((Integer 0) (Integer 1)))
	        (n (((. int mul) n) (factorial (((. int sub) n) (Integer 1)))))))
	      n)))))
	 (factorial (Integer 5)))

inferred type:

	int

Runtime term:

	(LetRec
	 (((1
	    (Lam
	     (1 ()
	      (App
	       ((Match
	         (BConst (cm_cases (((Integer 0) (Const (Integer 1)))))
	          (cm_default
	           ((Lam
	             (2 ()
	              (App
	               ((App
	                 ((App
	                   ((GetField mul)
	                    (Record
	                     ((add (Prim <fun>)) (compare (Prim <fun>))
	                      (div (Prim <fun>)) (mul (Prim <fun>))
	                      (rem (Prim <fun>)) (sub (Prim <fun>))))))
	                  (Var 0)))
	                (App
	                 ((Var 2)
	                  (App
	                   ((App
	                     ((App
	                       ((GetField sub)
	                        (Record
	                         ((add (Prim <fun>)) (compare (Prim <fun>))
	                          (div (Prim <fun>)) (mul (Prim <fun>))
	                          (rem (Prim <fun>)) (sub (Prim <fun>))))))
	                      (Var 0)))
	                    (Const (Integer 1))))))))))))))
	        (Var 0)))))))
	  (App ((Var 0) (Const (Integer 5))))))

Evaluated:

	(Const (Integer 120))
